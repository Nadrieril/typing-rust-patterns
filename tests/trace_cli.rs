use assert_cmd::Command;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[test]
fn trace_cli() -> anyhow::Result<()> {
    let test_cases = &[
        "quit",
        "help",
        "options",
        "set",
        "set mistyped",
        "set mistyped val",
        "set ref_binding_on_inherited true",
        "set allow_ref_pat_on_ref_mut false",
        "set downgrade_shared_inside_shared true",
        "set ergo2024\noptions",
        "&x: &mut T",
        "set allow_ref_pat_on_ref_mut false\n&x: &mut T",
        "rules",
    ];
    for &test_case in test_cases {
        let success = Command::cargo_bin("typing-rust-patterns")?
            .write_stdin(test_case)
            .assert()
            .success();
        let out = String::from_utf8(success.get_output().stdout.clone())?;

        // Identify each snapshot file by the hash of the test case.
        let req_hash = {
            let mut hasher = DefaultHasher::new();
            test_case.hash(&mut hasher);
            hasher.finish().to_string()
        };
        insta::with_settings!({
            snapshot_suffix => req_hash,
            info => &test_case,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(out);
        });
    }
    Ok(())
}
