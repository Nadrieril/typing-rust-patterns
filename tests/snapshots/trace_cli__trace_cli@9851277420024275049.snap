---
source: tests/trace_cli.rs
info: "set stable_rust\nrules"
---
allow_ref_pat_on_ref_mut: true -> false
eat_inherited_ref_alone: true -> false
inherited_ref_on_ref: "EatOuter" -> "EatBoth"
mut_binding_on_inherited: "Keep" -> "ResetBindingMode"
ref_binding_on_inherited: "Error" -> "Skip"
simplify_expressions: true -> false
The current options can be fully described as the following set of rules.

p0 @ p.0: T0, p1 @ p.1: T1
-------------------------- "Constructor"
[p0, p1] @ p: [T0, T1]

p0 @ &(*p).0: &T0, p1 @ &(*p).1: &T1
------------------------------------ "ConstructorRef"
[p0, p1] @ p: &[T0, T1]

p0 @ &mut (*p).0: &mut T0, p1 @ &mut (*p).1: &mut T1
---------------------------------------------------- "ConstructorRef"
[p0, p1] @ p: &mut [T0, T1]

[p0, p1] @ *p: &T
----------------- "ConstructorMultiRef"
[p0, p1] @ p: &&T

[p0, p1] @ &**p: &T
--------------------- "ConstructorMultiRef"
[p0, p1] @ p: &&mut T

[p0, p1] @ *p: &T
--------------------- "ConstructorMultiRef"
[p0, p1] @ p: &mut &T

[p0, p1] @ &mut **p: &mut T
--------------------------- "ConstructorMultiRef"
[p0, p1] @ p: &mut &mut T

p @ *p: T
---------------------------------- "Deref(EatOuter)"
&p @ p: &T, binding_mode(p) = move

p @ *p: T
------------------------------------------ "Deref(EatOuter)"
&mut p @ p: &mut T, binding_mode(p) = move

p @ *p: T
------------ "Deref(EatBoth)"
&p @ &p: &&T

p @ *p: T
-------------------- "Deref(EatBoth)"
&p @ &mut p: &mut &T

p @ *p: T
-------------------- "Deref(EatBoth)"
&mut p @ &p: &&mut T

p @ *p: T
---------------------------- "Deref(EatBoth)"
&mut p @ &mut p: &mut &mut T

x @ &p: &T
-------------- "BindingOverrideBorrow"
ref x @ &p: &T

x @ &p: &T
---------------------- "BindingOverrideBorrow"
ref x @ &mut p: &mut T

x @ &mut p: &mut T
------------------ "BindingOverrideBorrow"
ref mut x @ &p: &T

x @ &mut p: &mut T
-------------------------- "BindingOverrideBorrow"
ref mut x @ &mut p: &mut T

mut x @ p: T
-------------- "BindingResetBindingMode"
mut x @ &p: &T

mut x @ p: T
---------------------- "BindingResetBindingMode"
mut x @ &mut p: &mut T

x @ &p: &T
------------------------------------ "BindingBorrow"
ref x @ p: T, binding_mode(p) = move

x @ &mut p: &mut T
---------------------------------------- "BindingBorrow"
ref mut x @ p: T, binding_mode(p) = move

let x: T = p
------------ "Binding"
x @ p: T

let mut x: T = p
------------------------------------ "Binding"
mut x @ p: T, binding_mode(p) = move
